### ISSUE TITLE
[JOSS review] Some fixes and suggestions for the documentation

### ISSUE THREAD

**Issue by mtremmel** [SWIFTSIM/swiftsimio#67]
JOSS review issue: openjournals/joss-reviews#2430

<h1> Summary </h1>

This is a really great code. There is a clearly demonstrated need and the authors of the code have found creative ways to allow the user to analyze large simulation data in a memory efficient way. The code is intuitive to use, as well. From a practical standpoint, I would have liked to see the code be able to interface more seamlessly with halo catalogs, but I do not think that is a huge issue. The paper is well written, as is the very detailed documentation. Below I outline a few suggestions and edits to be made to the documentation. In particular, I had issues performing some functions and had to update a package (numba).

<h1> Documentation Comments </h1>

1) Given that the typical user will likely be interacting with a halo catalog, it would be nice to have a section devoted to such examples. Currently, direct use of velociraptor catalogs is deferred to the section on rotating projections. A clear example of how one could use his code in conjunction with a halo catalog will be useful. Making a separate section would make this information easier to find. At the start of this section would also be a useful place to link to the velociraptor code repository as well. This would likely be a good fit near the beginning of the Masking section.

2) In the Projections:Backends sub-section, when discussing sub-sampled projections the example given uses the star formation rate but for this particular example dataset this value is all 0. It might be nice to present an example that can actually produce a meaningful image. 

3) In the Projection:Rotation subsection, there is a typo. The variable `vector` passed as an argument to `rotation_matrix_from_vector()` should be `angular_momentum_vector` I think.

4) Also in the Rotation subsection, I was initially unable to run the line `un_rotated = project_gas_pixel_grid(**common_arguments)` and instead got the traceback below. The problem was solved by updating numba from 0.48.0 to 0.50.1 (I had to update my entire anaconda environment to do this, but based on the traceback I think numba was the key). It should be made clear that there is a version dependence on numba.

```
In [27]: un_rotated = project_gas_pixel_grid(**common_arguments) 
    ...:                                                                                                                                                   
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
<ipython-input-27-73ddb1693c8e> in <module>
----> 1 un_rotated = project_gas_pixel_grid(**common_arguments)

~/anaconda3/lib/python3.6/site-packages/swiftsimio/visualisation/projection.py in project_gas_pixel_grid(data, resolution, project, region, mask, rotation_matrix, rotation_center, parallel, backend)
    303         rotation_matrix=rotation_matrix,
    304         rotation_center=rotation_center,
--> 305         backend=backend,
    306     )
    307 

~/anaconda3/lib/python3.6/site-packages/swiftsimio/visualisation/projection.py in project_pixel_grid(data, boxsize, resolution, project, region, mask, rotation_matrix, rotation_center, parallel, backend)
    139             )
    140 
--> 141     number_of_particles = data.coordinates.shape[0]
    142 
    143     if project is None:

~/anaconda3/lib/python3.6/site-packages/swiftsimio/reader.py in getter(self)
    835                                     output_shape=output_shape,
    836                                     output_type=output_type,
--> 837                                     columns=columns,
    838                                 ),
    839                                 unit,

~/anaconda3/lib/python3.6/site-packages/swiftsimio/accelerated.py in read_ranges_from_file(handle, ranges, output_shape, output_type, columns)
    495     )
    496 
--> 497     return read_ranges(handle, ranges, output_shape, output_type, columns)

~/anaconda3/lib/python3.6/site-packages/swiftsimio/accelerated.py in read_ranges_from_file_chunked(handle, ranges, output_shape, output_type, columns)
    397 
    398     # Make array of chunk ranges
--> 399     chunk_ranges = get_chunk_ranges(ranges, chunk_size, handle.shape[0])
    400     chunk_range_size = np.diff(chunk_ranges).sum()
    401 

~/anaconda3/lib/python3.6/site-packages/numba/dispatcher.py in _compile_for_args(self, *args, **kws)
    399                 e.patch_message(msg)
    400 
--> 401             error_rewrite(e, 'typing')
    402         except errors.UnsupportedError as e:
    403             # Something unsupported is present in the user code, add help info

~/anaconda3/lib/python3.6/site-packages/numba/dispatcher.py in error_rewrite(e, issue_type)
    342                 raise e
    343             else:
--> 344                 reraise(type(e), e, None)
    345 
    346         argtypes = []

~/anaconda3/lib/python3.6/site-packages/numba/six.py in reraise(tp, value, tb)
    666             value = tp()
    667         if value.__traceback__ is not tb:
--> 668             raise value.with_traceback(tb)
    669         raise value
    670 

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Direct iteration is not supported for arrays with dimension > 1. Try using indexing instead.
[1] During: typing of intrinsic-call at /home/michael/anaconda3/lib/python3.6/site-packages/swiftsimio/accelerated.py (237)

File "anaconda3/lib/python3.6/site-packages/swiftsimio/accelerated.py", line 237:
def get_chunk_ranges(
    <source elided>
    n_ranges = len(ranges)
    for bounds in ranges:
    ^
```

5) In the Py-SPHViewer section, I believe there is a typo. When you call `SPHViewer()` in this code snippet I think you want to call `SPHViewerWrapper()`. 

6) Also in the Py-SPHViewer section, just for completeness include an `import matplotlib.pyplot as plt` in the code snippet.

**Comment by JBorrow** [comment]
Hi, thanks for your kind words and detailed review! I'm especially glad that you found the API intuitive, as that was the major goal when it came to the design.

To address the points in your review:

1. I have added a whole section on this, which loads a halo from a catalogue and makes images of it. Hopefully the inline images give people a good idea of what's going on, too. The reason this wasn't done before is that I am not fully happy with the VR library API so it's likely to change in the near future, but you're right that this needed more documentation. That's available here: https://swiftsimio.readthedocs.io/en/latest/velociraptor/index.html.

2. Fixed, I've changed this to `entropies`.

3. Thanks for the catch, I have fixed that.

4. This is a tricky one; I have been trying for a long while to not require `numba` on install, but now I think we've crossed the threshold where it's used too much throughout the code that it really needs to be there. Alongside this, the `numba` API is not very stable, so we fix `swiftsimio` when they update things but then lose backwards compatibility. I have now required `numba>=0.50.0`.

5. Thanks, fixed.

6. Thanks, fixed.

Thanks again for taking the time to do this review! I will push out v4.0.3 with your comments addressed.
